src/hip_stream.cpp

```diff
@@ -257,11 +257,85 @@ hipError_t hipStreamGetPriority(hipStream_t stream, int* priority) {
 hipError_t hipStreamAddCallback(hipStream_t stream, hipStreamCallback_t callback, void* userData,
                                 unsigned int flags) {
     HIP_INIT_API(hipStreamAddCallback, stream, callback, userData, flags);
-    hipError_t e = hipSuccess;
 
-    // Create a thread in detached mode to handle callback
-    ihipStreamCallback_t* cb = new ihipStreamCallback_t(stream, callback, userData);
-    std::thread(ihipStreamCallbackHandler, cb).detach();
+    auto stream_original{stream};
+    stream = ihipSyncAndResolveStream(stream);
+
+    if (!stream) return hipErrorInvalidValue;
+
+    LockedAccessor_StreamCrit_t cs{stream->criticalData()};
+
+    // stream is locked, now lock underlying HSA queue using unique_ptr RAII
+    auto av{cs->_av};
+    auto d{[=](hsa_queue_t*) mutable { av.release_locked_hsa_queue(); }};
+    std::unique_ptr<hsa_queue_t, decltype(d)> q{
+        static_cast<hsa_queue_t*>(av.acquire_locked_hsa_queue()),
+        std::move(d)};
+
+    // get first packet
+    auto b_index{hsa_queue_add_write_index_relaxed(q.get(), 1) % q->size};
+    auto b{static_cast<hsa_barrier_or_packet_t*>(q->base_address) + b_index};
```

Comment:
nit: not sure about single letter variable identifiers 'b' and 'd'. Maybe a more meaningful/longer identifier, making it easier to spot it in code, improve readability?